File: 1501-2000/1539_findKthPositive.py
class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        last = 0
        length = len(arr)
        left, right = 0, length - 1
        missing = 0
        while left <= right:
            mid = left + (right - left) // 2
            count = mid - left + 1
            elements = arr[mid] - last - count
            if missing + elements < k:
                missing += elements
                last = arr[mid]
                left = mid + 1
            else:
                right = mid - 1
        return last + k - missing

# Alternative solution

class Solution:
    def findKthPositive(self, arr: List[int], k: int) -> int:
        i = 0
        num = 0

        while i < len(arr) and k:
            if arr[i] == num + 1:
                i += 1
            else:
                k -= 1
            num += 1

        return num + k

File: 1501-2000/1539_findKthPositive.java
class Solution {
    public int findKthPositive(int[] arr, int k) {
        return binarySearch(arr, k, 0, 0, arr.length - 1);
    }

    public int binarySearch(int[] arr, int k, int base, int start, int end) {
        if (end < start) {
            return (end >= 0) ? arr[end] + k : k;
        }

        int mid = start + (end - start) / 2;
        int num = arr[mid];
        int count = num - base - (mid - start + 1);
        int newK = k - count;

        if (newK > 0) {
            return binarySearch(arr, newK, arr[mid], mid + 1, end);
        } else {
            return binarySearch(arr, k, base, start, end - 1);
        }
    }
}
